haystack_experimental/components/agents/agent.py

```diff
@@ -2,29 +2,62 @@
 #
 # SPDX-License-Identifier: Apache-2.0
 
-import inspect
+from dataclasses import dataclass
 from typing import Any, Optional, Union
 
+# Monkey patch Haystack's AgentSnapshot with our extended version
+import haystack.dataclasses.breakpoints as hdb
+
+from haystack_experimental.dataclasses.breakpoints import AgentSnapshot
+
+# replace reference
+hdb.AgentSnapshot = AgentSnapshot
+
 from haystack import logging
 from haystack.components.agents.agent import Agent as HaystackAgent
-from haystack.components.agents.state.state import _schema_from_dict, _schema_to_dict, _validate_schema
-from haystack.components.agents.state.state_utils import merge_lists
+from haystack.components.agents.agent import _ExecutionContext as Haystack_ExecutionContext
+from haystack.components.agents.agent import _schema_from_dict
 from haystack.components.generators.chat.types import ChatGenerator
-from haystack.core.component.component import component
+from haystack.core.errors import PipelineRuntimeError
+from haystack.core.pipeline import AsyncPipeline, Pipeline
+from haystack.core.pipeline.breakpoint import (
+    _create_pipeline_snapshot_from_chat_generator,
+    _trigger_tool_invoker_breakpoint,
+)
 from haystack.core.pipeline.utils import _deepcopy_with_exceptions
-from haystack.core.serialization import component_to_dict, default_from_dict, default_to_dict, import_class_by_name
+from haystack.core.serialization import default_from_dict, import_class_by_name
 from haystack.dataclasses import ChatMessage
+from haystack.dataclasses.breakpoints import AgentBreakpoint, PipelineSnapshot, ToolBreakpoint
 from haystack.dataclasses.streaming_chunk import StreamingCallbackT
-from haystack.tools import Tool, Toolset, deserialize_tools_or_toolset_inplace, serialize_tools_or_toolset
-from haystack.utils.callable_serialization import deserialize_callable, serialize_callable
+from haystack.tools import Tool, Toolset, deserialize_tools_or_toolset_inplace
+from haystack.utils.callable_serialization import deserialize_callable
 from haystack.utils.deserialization import deserialize_chatgenerator_inplace
 
-from haystack_experimental.components.agents.human_in_the_loop.types import ConfirmationStrategy
-from haystack_experimental.components.tools.tool_invoker import ToolInvoker
+from haystack_experimental.components.agents.human_in_the_loop import (
+    ConfirmationStrategy,
+    ToolExecutionDecision,
+    ToolBreakpointException
+)
+from haystack_experimental.components.agents.human_in_the_loop.strategies import _handle_confirmation_strategies
+from haystack_experimental.core.pipeline.breakpoint import _create_pipeline_snapshot_from_tool_invoker
 
 logger = logging.getLogger(__name__)
 
 
+@dataclass
+class _ExecutionContext(Haystack_ExecutionContext):
+    """
+    Execution context for the Agent component
+
+    Extends Haystack's _ExecutionContext to include tool execution decisions for human-in-the-loop strategies.
```

Comment:
`_ExecutionContext` is extended to include `tool_execution_decisions`. This was necessary to get the `BreakpointConfirmationStrategy` to work since we needed some way to input the `tool_execution_decisions` that may have come externally.